### PR TITLE
[ENH] Add optimizer param for neuralforecast models

### PR DESCRIPTION
**Pull request by pranavvp16**
As discussed on discord I have updated the neuralforecast models to support the optimizer parameter. Just a question to @jmoralez if I don't pass the `optimizer` and pass the `optimizer_kwargs` will it overide the kwargs of default optimizer or just ignore the kwargs, as it doesn't raise error when `optimizer_kwargs` are passed without `optimizer`.

### CODE REVIEW

**Comment by jmoralez** [comment]
Hey, it'll ignore them. We can add a warning for that case.

**Comment by pranavvp16** [comment]
@fkiraly  the tests are failing due to addition of new dependanacy which is torch. NeuralForecast has added new feature to support custom optimiser to neuralforecast models which have to be from class torch.optim.Optimiser. I thought that neuralforecast has the dependancy on torch and torch will be downloaded along the way but the tests fail with module not found error. Should I add torch to python-dependancies list in the tags of forecaster to pass the tests ??

**Comment by yarnabrina** [comment]
`torch` was there already in `dl` dependency where `neuralforecast` is there. It's failing now as you are trying to import at the module level, which expects it to be present always which is neither the case nor desirable.

To fix, you should move the import elsewhere, and remove the top level type hint. Please take a look at how I handled loss specification.

Also, you must raise the upper bound on `neuralforecast` for CI to pass, otherwise I think it's going to fail. We should not upgrade lower bound though, so you need to handle for the cases where users pass arguments that are only present in higher versions. Please take a look at `statsforecast` adapter where @arnaujc91 added such adaptations.

**Comment by pranavvp16** [comment]
So basically we need to handle the model parameters for the latest version as well as version `1.6.4` right ?? @yarnabrina

File: sktime/forecasting/neuralforecast.py
Comment (yarnabrina) [inline sktime/forecasting/neuralforecast.py:164]: This is a potentially breaking change. Users who are using neuralforecast 1.6.4 with sktime currently will not longer be able to use this estimator any more. Especially if they don't want to customose optimiser, it will be unnecessary. Please revert this change.

File: sktime/forecasting/neuralforecast.py
Comment (yarnabrina) [inline sktime/forecasting/neuralforecast.py:529]: Same as above.

File: sktime/forecasting/neuralforecast.py
Comment (yarnabrina) [inline sktime/forecasting/neuralforecast.py:200]: New arguments should be added at end of function, otherwise it may be a breaking change for users who uses keyword arguments as positional.

File: sktime/forecasting/neuralforecast.py
Comment (yarnabrina) [inline sktime/forecasting/neuralforecast.py:564]: Same as above.

**Review by yarnabrina — CHANGES_REQUESTED** [review]
Please take a look at https://github.com/sktime/sktime/pull/5112 and related changes. In `_GeneralisedStatsForecastAdapter`, we have a method to check user provided arguments through sktime and a method to identify supported arguments based on installed statsforecast version. Only the supported ones are used, and the rest is ignored with an warning. We need to do the similar thing here as well.

Further more, your current changes are not tested in CI. You need to update `pyproject.toml`. Currently it uses `'neuralforecast<1.7.0,>=1.6.4; python_version < "3.11"'`, and we should modify it to support latest release. Assuming `neuralforecast` follows semantic versioning, we can update it to `'neuralforecast<1.8.0,>=1.6.4; python_version < "3.11"'`. The python version bound will remain as it is, as `neuralforecast` do not support `>3.10` officially yet.

Please let me know if you have any doubts/questions.

---

A small comment: you don't need to add test for every single addition of new argument. Of course it adds more coverage, but the it's not scalable in long run with more and more arguments supported by underlying packages. We trust they have their own unit tests so we will rely on that. This is my personal opinion by the way, may be @fkiraly, @benHeid can comment if they agree or not. Either way, you do not have to remove your existing codes, it's just a comment for future additions.

**Comment by fkiraly** [comment]
> This is my personal opinion by the way, may be @fkiraly, @benHeid can comment if they agree or not. 

In-principle yes - we should make sure though that new parameters are covered, so we notice when they are removed. My method of choice would be via `get_test_params`, but I don't know enough of the details to guess how robust this is for this specific case.

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_neuralforecast.py:7]: I don't think this is correct. You should throw an actual warning, using `warnings` module or the `sktime` wrapper for the same instead of logging a message at WARNING level.

Use `from sktime.utils.warnings import warn` and change accordingly.

Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_neuralforecast.py:7]: Also, please keep the imports sorted. You can use tools like `isort`.

Optional but highly recommended.

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_neuralforecast.py:148]: I am not sure if I follow this. Why are you checking for parameters of `Trainer` class from `pytorch-lightning`? Should you not check only for the underlying model class in `neuralforecast`? Can you please explain?

Does `neuralforecast` really accept all parameters of `Trainer` to be passed directly? If for example someone modifies `__init__` of `NeuralForecastLSTM` class with every single argument of `Trainer` clas, those will pass this check but eventually fail right? Please let me know if I am missing something obvious.

**Review by yarnabrina — CHANGES_REQUESTED** [review]
I mdae a few comments, please take a look.

I will also consider this PR to be dependent of the PR #6237 by @geetu040 , as that fixes a bug in the adapter introduced by new release, and then this PR adds new enhancements.

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (pranavvp16) [reply sktime/forecasting/base/adapters/_neuralforecast.py:148]: yes yes neuralforecast models have a parameter like **trainer_kwargs which accepts all the parameter for lightning [trainer](https://nixtlaverse.nixtla.io/neuralforecast/models.lstm.html)

Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_neuralforecast.py:148]: Thanks for pointing that out. I completely forgot.

**Comment by pranavvp16** [comment]
Fixed the changes suggested replaced the logging.warn with sktime util. Will wait for #6237 to get merged, fix the merge conflicts and get this merged.

**Comment by yarnabrina** [comment]
@pranavvp16 #6237 is merged now, let's resume this PR once again.

After this is merged as well, I'll resume #6124 .

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (VascoSch92) [inline sktime/forecasting/base/adapters/_neuralforecast.py:162]:
```suggestion
                        f"from {self.algorithm_class}. \n"
```
to make the message ore readable

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (VascoSch92) [inline sktime/forecasting/base/adapters/_neuralforecast.py:152]: just a random question: can we have duplicate? if yes, this can be a problem or?

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_neuralforecast.py:158]: Can you please explain why "not None" is required, and why should we not warn always?

Are you intending this to be a check for non-default value passed by users? Does all parameters have None as default?

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_neuralforecast.py:166]: I'll suggest to avoid using pop as it's an in-place modification and moat liky modify self.algorithm_parameters too. Will it make sense to do a deep copy in a new variable first, or maintain a separate dictionary with only valid parameters?

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_neuralforecast.py:176]: Typo in comment: filer -> filter?

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_neuralforecast.py:163]: Question:  will it make sense to raise different warnings for discrepancy with neuralforecast or with pytorch lightning?

File: sktime/forecasting/neuralforecast.py
Comment (yarnabrina) [inline sktime/forecasting/neuralforecast.py:100]: Can you please add supported iptimisers and a link tobthe documentation in reference section? 

Same comment applies for other models too.

**Review by yarnabrina — CHANGES_REQUESTED** [review]

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (pranavvp16) [reply sktime/forecasting/base/adapters/_neuralforecast.py:158]: if suppose the user is on version `1.6.4` and didn't pass the optimizer still he will get a warning that optimizer is being dropped as top level interfaces pass optimizer param through `self.algorithm_params`. Now the default value of optimizer is None so it will have no effect on the model, but suppose the value of optmizer by is passed by user or is optim object by default  then showing a warning that it is being dropped would be good as it will have effect on the model, while params with None won't have any effect. So basically this will log warning only when user passes a invalid parameter with actual value or the default value of the parameter will affect the model

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (pranavvp16) [reply sktime/forecasting/base/adapters/_neuralforecast.py:166]: yes I think it would be good to make a deepcopy of filter params

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (pranavvp16) [reply sktime/forecasting/base/adapters/_neuralforecast.py:163]: We have issues with params of neuralforecast, so why raise warnings for pytorch lightning ?? I'm unable to understand what are you trying to say here

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_neuralforecast.py]: > Are you intending this to be a check for non-default value passed by users? Does all parameters have None as default?

Are you certain that all newly added parameters in `neuralforecast` in future releases will have default value of `None`? Just for illustration purposes, what happens if they add `n_jobs=-1`?

Will it not be better to compare against the default values of sktime, e.g. using `self.get_default_params`? I'll suggest to check the `statsforecast` changes by @arnaujc91 for reference.

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_neuralforecast.py]: Suppose neuralforecast adds upper  bound for pytorch lightning to version X, where X changes over time. And suppose Y is the latest version of pytorch lightning, which also vary over time.

If Y has few parameters that X lacks, and user sees them from the website and passes them to trainer arguments, it will fail on sktime main as those are unsupported. If I understand correctly, your changes will be able to capture and drop those extra ones, but the warning will inform users about neuralforecast version limitation. But these warnings will originate due to pytorch lightning version discrepancies, not neuralforecast. 

Let me know if it makes sense.

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (pranavvp16) [reply sktime/forecasting/base/adapters/_neuralforecast.py]: Yes it will be to better to compare with default values

File: sktime/forecasting/base/adapters/_neuralforecast.py
Comment (pranavvp16) [reply sktime/forecasting/base/adapters/_neuralforecast.py]: I don't expect a class like `Trainer` to add a new param in further versions or update. Still if you think adding a warning for params not supported due pytorch_ligthning version I will add that. But this won't happen most probably. Let me know what do you think about it ??

**Comment by yarnabrina** [comment]
I see a lot of jobs were cancelled, do you know why? Can you please rebase once from main (if not done already), and trigger the tests again?

**Comment by fkiraly** [comment]
> I see a lot of jobs were cancelled, do you know why? Can you please rebase once from main (if not done already), and trigger the tests again?

That might have been me, I cancelled the PRs prior to the fix of the failures on `main` in #6280, and probably forgot to restart this one.

**Review by yarnabrina — CHANGES_REQUESTED** [review]
I added few minor comments, please address those and we can merge this PR then today before tomorrow's scheduled full run (UTC 12 on Sunday) on `main`.

Other than the below concepts, I've no more crucial feedbacks, except few very minor changes. Those can be done in a separate PR afterwards. Unless @fkiraly or @jmoralez have any other concerns, I'll be happy to merge after the following changes.

**Review by yarnabrina — APPROVED** [review]

**Pull request merged**
